lib/opte/src/engine/snat.rs

```diff
@@ -177,129 +186,116 @@ impl<T: ConcreteIpAddr> FiniteResource for NatPool<T> {
 
 /// A NAT pool mapping provided for Source NAT (only outbound connections).
 #[derive(Clone)]
-pub struct SNat {
-    priv_ip: Ipv4Addr,
-    ip_pool: Arc<NatPool<Ipv4Addr>>,
+pub struct SNat<T: ConcreteIpAddr> {
+    priv_ip: T,
+    ip_pool: Arc<NatPool<T>>,
+}
+
+enum GenIcmpErr<T: Display> {
+    MetaNotFound,
+    NotRequest(T),
+}
+
+impl<T: Display> From<GenIcmpErr<T>> for GenDescError {
+    fn from(val: GenIcmpErr<T>) -> Self {
+        GenDescError::Unexpected {
+            msg: match val {
+                GenIcmpErr::MetaNotFound => {
+                    "No ICMP metadata found despite Protocol::ICMP".to_string()
+                }
+                GenIcmpErr::NotRequest(v) => {
+                    format!("Expected ICMP Echo Request, found: {}", v)
+                }
+            },
+        }
+    }
 }
 
-impl SNat {
-    pub fn new(addr: Ipv4Addr, ip_pool: Arc<NatPool<Ipv4Addr>>) -> Self {
+impl<T: ConcreteIpAddr + 'static> SNat<T> {
+    pub fn new(addr: T, ip_pool: Arc<NatPool<T>>) -> Self {
         SNat { priv_ip: addr, ip_pool }
     }
 
-    // A helper method for generating an SNAT + ICMP action descriptor.
+    // A helper method for generating an SNAT + ICMP(v6) action descriptor.
     fn gen_icmp_desc(
         &self,
-        nat: NatPoolEntry<Ipv4Addr>,
+        nat: NatPoolEntry<T>,
         pkt: &Packet<Parsed>,
     ) -> GenDescResult {
         let meta = pkt.meta();
 
-        if let Some(icmp) = meta.inner_icmp() {
-            if icmp.msg_type.inner != Icmpv4Message::EchoRequest {
-                return Err(GenDescError::Unexpected {
-                    msg: format!(
-                        "Expected ICMP Echo Request, found: {}",
-                        icmp.msg_type,
-                    ),
-                });
+        let echo_ident = match T::MESSAGE_PROTOCOL {
+            Protocol::ICMP => {
+                let icmp = meta
+                    .inner_icmp()
+                    .ok_or(GenIcmpErr::<Icmpv4Message>::MetaNotFound)?;
+                if icmp.msg_type != Icmpv4Message::EchoRequest.into() {
+                    Err(GenIcmpErr::NotRequest(icmp.msg_type))?;
```

Comment:
Nit: I would just use a simple return instead of `?` on an `Err` here (and below) or alternative also use `Ok(icmp.echo_id())` and apply `?` to the entire match.